Issue #1278

evilmopey opened the pull request:
Created a compensate_lock function that make sure the right amount of balls are in the lock. If the numbers don't match up the machine will compensate by removing or adding balls.

evilmopey committed:
Merge pull request #1 from missionpinball/dev

Try

evilmopey committed:
Update multiball_lock.py

issue 1278

avanwinkle reviewed (commented):
How will this affect machines that don't have a path directly to the lock?

jabdoa2 commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +305,26 @@ def _post_events(self, device, **kwargs):
+            self._locked_balls = 0
+            # remove balls that were in lock
+            balance = (self._locked_balls - (-1 * self._physically_locked_balls))  # -1 to make the number positive
+            for i in range(1, balance):
```
I don't think we need to remove balls. That would give the current player more balls.

avanwinkle commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +305,26 @@ def _post_events(self, device, **kwargs):
+    
+    def compensate_lock(self):
+        # evilmopey solution for issue 1278
+        if self.config['locked_ball_counting_strategy'] == "no_virtual":
```
If this approach is only applicable to the `"no_virtual"` approach, could you wrap the calling function above instead of having so many conditionals?

```
if self.config['locked_ball_counting_strategy'] == "no_virtual":
  self.compensate_lock()
```

avanwinkle commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +305,26 @@ def _post_events(self, device, **kwargs):
+            # This means we need to remove balls from the physical lock
+            balance = self._physically_locked_balls - self._locked_balls
+            for i in range(1, balance):
+                self.lock_devices[0].eject()
```
From my understanding, with no virtual lock counting then any balls locked by player 1 will be ejected before player 2 starts. If that's the case, you'll want to block the queue for these balls to drain so they're not on the playfield when player 2's turn starts. See above for an example

jabdoa2 commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +305,26 @@ def _post_events(self, device, **kwargs):
+        if self._locked_balls > self._physically_locked_balls:
+             # This means we need to add more balls to the physical lock
+            balance = self._locked_balls - self._physically_locked_balls
+            self._request_new_balls(balance)
```
The condition looks right. However, request_new_balls is not the correct function because it will request the ball to the playfield. Instead you would call request_ball on self.lock_devices[x].

evilmopey replied:
So I took out the if statements for both no_virtual and physical_only leaving only the if statement that pertains to virtual_only or min_virtual_physical. This if statement only adds balls not takes them away. My only problem now is that on line 9 MYPY is false, this blocks the class BALLDevice from being imported and being able to use the request_ball function. I added the import above it so I could use the request_ball function.

jabdoa2 commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +305,26 @@ def _post_events(self, device, **kwargs):
+            # This means we need to remove balls from the physical lock
+            balance = self._physically_locked_balls - self._locked_balls
+            for i in range(1, balance):
+                self.lock_devices[0].eject()
```
Removing balls is not needed I guess. This would require some more logic.

evilmopey replied:
Thanks for the feedback everyone I will edit my code with your input in mind.

jabdoa2 reviewed (changes requested):
The approach is good. Some small comments below.

I think it would help to add a test case to see if things work as expected.

With the test we will also need to check if there actually a path to the trough. Otherwise some other tests will fail or the code will crash.

evilmopey committed:
Update multiball_lock.py

I shortened the function to only test for virtual_only and min_virtual_physical and removed any part of the function that took away balls from the lock. The only problem I am running into is that MYPY is declared false so BallDevice is not imported that means that the function request_ball cannot be called. I added an import statement above it so that there were no errors.

avanwinkle commented on mpf/devices/multiball_lock.py:
```diff
@@ -305,3 +306,12 @@ def _post_events(self, device, **kwargs):
+            if self._locked_balls > self._physically_locked_balls:
+                # This means we need to add more balls to the physical lock
+                ball_deficit = self._locked_balls - self._physically_locked_balls
+                BallDevice.request_ball(self.lock_devices[0], ball_deficit)
```
How does `request_ball` work for locks that don't have a route from the trough? Will it be ignored, or wait until a playfield ball lands in it?

evilmopey replied:
So when request_ball is called it starts a for loop that iterates as many times as the value balls were passed in. Each iteration calls setup_queue_eject_to_target, if the ball cannot be routed to the target a tuple entry will then be made into the _ball_requests_ deque. So to my knowledge the request is not being ignored.

evilmopey commented:
So what would a test case look like on this project?

jabdoa2 commented:
> So what would a test case look like on this project?

You could extend this test: https://github.com/missionpinball/mpf/blob/dev/mpf/tests/test_MultiballLock.py. The corresponding config would be here: https://github.com/missionpinball/mpf/tree/dev/mpf/tests/machine_files/multiball_locks. You can use self.start_game()+self.add_player() to start a two player game. Then you can use self.drain_one_ball() to end the ball.

evilmopey commented:
I am trying to test the issue but keep getting errors, can someone point out what I am doing wrong?
https://github.com/evilmopey/issue-1278
